File: android/src/com/sanztech/qrisgate/QRISNotificationListener.java
package com.sanztech.qrisgate;

import android.service.notification.NotificationListenerService;
import android.service.notification.StatusBarNotification;
import android.util.Log;
import android.content.SharedPreferences;
import java.net.HttpURLConnection;
import java.net.URL;
import java.io.OutputStream;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

public class QRISNotificationListener extends NotificationListenerService {
    private static final String TAG = "QRISListener";

    @Override
    public void onNotificationPosted(StatusBarNotification sbn) {
        String packageName = sbn.getPackageName();

        // Map package to readable source name
        String source = packageName;
        if (packageName.contains("dana")) source = "DANA";
        else if (packageName.contains("gojek") || packageName.contains("gopay")) source = "GoPay";
        else if (packageName.contains("ovo")) source = "OVO";
        else if (packageName.contains("shopee")) source = "ShopeePay";
        else if (packageName.contains("bca")) source = "BCA";
        else if (packageName.contains("linkaja")) source = "LinkAja";
        else if (packageName.contains("bri")) source = "BRI";
        else if (packageName.contains("mandiri")) source = "Mandiri";
        else if (packageName.contains("bni")) source = "BNI";

        String title = "";
        String text = "";
        try {
            title = sbn.getNotification().extras.getString("android.title", "");
            text = sbn.getNotification().extras.getString("android.text", "");
        } catch (Exception e) {
            Log.e(TAG, "Error reading notification", e);
        }
        if (title == null) title = "";
        if (text == null) text = "";

        Log.d(TAG, "Notif from " + source + ": " + title + " - " + text);
        addLog("📩 " + source + ": " + title + " - " + text);

        // Check for payment keywords
        String combined = (title + " " + text).toLowerCase();
        if (combined.contains("berhasil") || combined.contains("sukses") ||
            combined.contains("pembayaran masuk") || combined.contains("diterima") ||
            combined.contains("transfer masuk") || combined.contains("uang masuk")) {

            // Extract nominal (all digits from text)
            String nominalStr = text.replaceAll("[^0-9]", "");
            if (nominalStr.isEmpty()) {
                nominalStr = title.replaceAll("[^0-9]", "");
            }

            if (!nominalStr.isEmpty()) {
                addLog("💰 Pembayaran terdeteksi! Rp " + formatRupiah(nominalStr) + " dari " + source);
                incrementNotifCount();
                sendToWebhook(nominalStr, source, title + " - " + text);
            }
        }
    }

    private void sendToWebhook(String nominal, String source, String rawText) {
        new Thread(() -> {
            try {
                SharedPreferences prefs = getSharedPreferences("qrisgate_prefs", MODE_PRIVATE);
                String webhookUrl = prefs.getString("webhook_url", "");
                String apiKey = prefs.getString("api_key", "");

                if (webhookUrl.isEmpty()) {
                    addLog("⚠️ Webhook URL belum diset!");
                    return;
                }

                URL url = new URL(webhookUrl);
                HttpURLConnection conn = (HttpURLConnection) url.openConnection();
                conn.setRequestMethod("POST");
                conn.setRequestProperty("Content-Type", "application/json");
                if (!apiKey.isEmpty()) {
                    conn.setRequestProperty("x-api-key", apiKey);
                    conn.setRequestProperty("Authorization", "Bearer " + apiKey);
                }
                conn.setConnectTimeout(15000);
                conn.setReadTimeout(15000);
                conn.setDoOutput(true);

                // Escape special characters in JSON values
                String safeSource = source.replace("\"", "\\\"");
                String safeRaw = rawText.replace("\"", "\\\"").replace("\n", "\\n");

                String json = "{\"nominal\":\"" + nominal + "\"," +
                              "\"source\":\"" + safeSource + "\"," +
                              "\"raw_text\":\"" + safeRaw + "\"}";

                try (OutputStream os = conn.getOutputStream()) {
                    os.write(json.getBytes("UTF-8"));
                }

                int responseCode = conn.getResponseCode();
                Log.d(TAG, "Webhook response: " + responseCode);
                addLog("✅ Webhook terkirim (HTTP " + responseCode + ")");
                conn.disconnect();
            } catch (Exception e) {
                Log.e(TAG, "Webhook error", e);
                addLog("❌ Webhook gagal: " + e.getMessage());
            }
        }).start();
    }

    private void addLog(String msg) {
        try {
            SharedPreferences prefs = getSharedPreferences("qrisgate_prefs", MODE_PRIVATE);
            String time = new SimpleDateFormat("HH:mm:ss", Locale.getDefault()).format(new Date());
            String entry = "[" + time + "] " + msg + "\n";
            String existing = prefs.getString("listener_logs", "");
            String newLogs = entry + existing;

            // Keep only last 2000 chars
            if (newLogs.length() > 2000) {
                newLogs = newLogs.substring(0, 2000);
            }

            prefs.edit().putString("listener_logs", newLogs).apply();
        } catch (Exception e) {
            Log.e(TAG, "Log error", e);
        }
    }

    private void incrementNotifCount() {
        SharedPreferences prefs = getSharedPreferences("qrisgate_prefs", MODE_PRIVATE);
        int count = prefs.getInt("notif_count", 0);
        prefs.edit().putInt("notif_count", count + 1).apply();
    }

    private String formatRupiah(String num) {
        try {
            long n = Long.parseLong(num);
            return String.format(Locale.getDefault(), "%,d", n).replace(',', '.');
        } catch (Exception e) {
            return num;
        }
    }
}

File: mobile/android/app/src/main/java/com/sanzceo/qrisgate/QRISNotificationListener.kt
package com.sanzceo.qrisgate

import android.service.notification.NotificationListenerService
import android.service.notification.StatusBarNotification
import android.util.Log
import android.content.Context
import java.net.HttpURLConnection
import java.net.URL
import java.io.OutputStream

class QRISNotificationListener : NotificationListenerService() {
    private val TAG = "QRISListener"

    override fun onNotificationPosted(sbn: StatusBarNotification) {
        val packageName = sbn.packageName
        
        var source = packageName
        if (packageName.contains("dana")) source = "dana"
        else if (packageName.contains("gojek") || packageName.contains("gopay")) source = "gopay merchant"
        else if (packageName.contains("ovo")) source = "ovo"
        else if (packageName.contains("shopee")) source = "shopeepay"
        else if (packageName.contains("bca")) source = "bca merchant"
        else if (packageName.contains("linkaja")) source = "linkaja"

        var title = sbn.notification.extras.getString("android.title", "")
        var text = sbn.notification.extras.getString("android.text", "")
        if (title == null) title = ""
        if (text == null) text = ""
        
        Log.d(TAG, "Notification received: $title - $text")

        val lowerText = text.lowercase()
        val lowerTitle = title.lowercase()

        if (lowerText.contains("berhasil") || lowerTitle.contains("berhasil") || 
            lowerText.contains("sukses") || lowerTitle.contains("sukses") || 
            lowerText.contains("pembayaran masuk")) {
            
            var nominalStr = text.replace("[^0-9]".toRegex(), "")
            if (nominalStr.isEmpty()) {
                nominalStr = title.replace("[^0-9]".toRegex(), "")
            }
            if (nominalStr.isNotEmpty()) {
                sendToWebhook(nominalStr, source)
            }
        }
    }

    private fun sendToWebhook(nominal: String, source: String) {
        Thread {
            try {
                val prefs = getSharedPreferences("qrisgate_prefs", Context.MODE_PRIVATE)
                val webhookUrl = prefs.getString("webhook_url", "https://qris-backend-mu.vercel.app/api/webhook")
                
                if (webhookUrl.isNullOrEmpty()) return@Thread

                val url = URL(webhookUrl)
                val conn = url.openConnection() as HttpURLConnection
                conn.requestMethod = "POST"
                conn.setRequestProperty("Content-Type", "application/json")
                conn.doOutput = true

                val jsonPayload = "{\"nominal\":\"$nominal\", \"source\":\"$source\"}"
                
                conn.outputStream.use { os ->
                    val input = jsonPayload.toByteArray(Charsets.UTF_8)
                    os.write(input, 0, input.size)
                }
                
                Log.d(TAG, "Webhook response code: ${conn.responseCode}")
                conn.disconnect()
            } catch (e: Exception) {
                Log.e(TAG, "Webhook error", e)
            }
        }.start()
    }
}
